### ISSUE TITLE
Getting Java::JavaLang::ArrayIndexOutOfBoundsException in compiler internals

### ISSUE THREAD

**Issue by fxposter** [jruby/jruby#6692]
**Environment Information**
  
Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
* Operating system and platform (e.g. `uname -a`)

Other relevant info you may wish to add:
* Installed or activated gems
* Application/framework version (e.g. Rails, Sinatra)
* Environment variables

**Expected Behavior**

* Describe your expectation of how JRuby should behave, perhaps by showing how CRuby/MRI behaves.
* Provide an executable Ruby script or a link to an example repository.

**Actual Behavior**

* Describe or show the actual behavior.
* Provide text or screen capture showing the behavior.

**Comment by headius** [comment]
Definitely some kind of IR issue, though unsure if it is systemic or an intermittent concurrent-access issue.

The index value reported is -2147480651, which is close to an overflowed 32-bit integer. Some bit masking got screwed up?

Are you able to narrow it down to a specific erb template? Perhaps we could try running it in a loop, with some concurrent loops, and trigger this.

@enebo Any quick theories on how this would happen at this point in the code?

**Comment by fxposter** [comment]
@headius No, I tried to reproduce on a smaller cases - and failed to do that. I don't control what erb's I evaluate (and not sure if it is related to erbs at all or surrounding code). Given the fact that I don't know when these internal closures are generated in the first place - I cannot come up with a reproduction scenario :(

**Comment by headius** [comment]
We will have a look at the code in the trace and try to come up with some theories.

**Comment by headius** [comment]
Looking at the code, this closureId seems to come from IRScope.getNextClosureId:

https://github.com/jruby/jruby/blob/e99987e1617105cce12ba2afa109d9944a297294/core/src/main/java/org/jruby/ir/IRScope.java#L823-L827

This is a signed integer, with a max value around 2.1 billion. The value in your error above is -2147480651, which is around 3000 greater than the overflowed signed integer value. Is it possible that your app evals its way through two billion closures/blocks before this happens?

**Comment by fxposter** [comment]
Yes, I saw that. It is possible (cause it can spend days before crashing), question is - what type of code triggers this getNextClosureId?

Also, the problem started to happen after upgrade to jruby 9.2 from 9.1. This upgrade was combined with a bunch of other changes, but it does seem that it was because of the jruby change version itself.

**Comment by headius** [comment]
The getClosureId is just used as a monotonic identifier for the closure body, for internal bookkeeping and logging. We could probably fix this by bumping the value up to a long, which should provide positive values until we are all dust.

@enebo Only impacts I can see from bumping this value would be the public API (not blessed for anyone to use) and IR serialization (but I have not confirmed it uses closure ID). I will put together a PR and see how it looks.

**Comment by enebo** [comment]
@headius @fxposter I think a long here is no big deal but are there actually billions of closures underneath a single scope?  I would expect you to have a verrrrrry big nestedClosures list if so.

**Comment by headius** [comment]
I have reproduced locally without blowing my memory up, so it seems this is the likely explanation.

```ruby
8.times.map {
  sleep 1
  Thread.new {
    i = 0
    while true
      i+=1
      puts Thread.current.inspect + ":" + i.to_s if i % 10_000_000 == 0
      eval "1.tap {}"
    end
  }
}.each(&:join)
```

**Comment by headius** [comment]
#6702 replaces the int with a long, but @enebo may have a more localized solution that doesn't use a single monotonic value for all closures (and evals, apparently).

**Comment by enebo** [comment]
So quick note.  I believe IREvalScript isa IRClosure so any new closures added to an eval looks below the eval for a new closure id.  This is why it overflows.

I am testing a different fix which makes evals a hard boundary for getting assigned a closure id.  This should not grow to crazy large numbers before each each ends up being a new IREvalScript instance.

**Comment by fxposter** [comment]
The problem is the fact that I don't `eval` in a context of a single object. the context is usually a new object all the time. maybe, there is some very long loop that `eval`-s somewhere, but I am not sure. That's why I didn't try to reproduce eval-ing within a single object.

**Comment by enebo** [comment]
@fxposter you can see above @headius is using  a simple form of eval.   So I think regardless of how the eval is called it is calling up past the eval to whatever lexical/syntactical scope is above the call and using its scope to give closures in the eval new ids.

**Comment by headius** [comment]
@fxposter We have merged the fix and will release it next week. Before then you could try out a snapshot build but I suppose there won't be time enough to confirm the bug is fixed. I am also confident that this fix will solve your issue.

**Comment by fxposter** [comment]
Thanks!
